File: modelling/generatedHaskellFiles/code/queue/Queue.hs
{-# LANGUAGE EmptyDataDecls, RankNTypes, ScopedTypeVariables #-}

module Queue(Queue, empty, dequeue, enqueue) where {

import Prelude ((==), (/=), (<), (<=), (>=), (>), (+), (-), (*), (/), (**),
  (>>=), (>>), (=<<), (&&), (||), (^), (^^), (.), ($), ($!), (++), (!!), Eq,
  error, id, return, not, fst, snd, map, filter, concat, concatMap, reverse,
  zip, null, takeWhile, dropWhile, all, any, Integer, negate, abs, divMod,
  String, Bool(True, False), Maybe(Nothing, Just));
import qualified Prelude;

data Queue a = AQueue [a] [a];

empty :: forall a. Queue a;
empty = AQueue [] [];

dequeue :: forall a. Queue a -> (Maybe a, Queue a);
dequeue (AQueue [] []) = (Nothing, AQueue [] []);
dequeue (AQueue xs (y : ys)) = (Just y, AQueue xs ys);
dequeue (AQueue (v : va) []) = (case reverse (v : va) of {
                                 y : ys -> (Just y, AQueue [] ys);
                               });

enqueue :: forall a. a -> Queue a -> Queue a;
enqueue x (AQueue xs ys) = AQueue (x : xs) ys;

}

File: modelling/queue.py
class Queue:
    empty = []

    def __init__(self) -> None:
        self.queue = []
    
    def enqueue(self, i):
        self.queue.append(i)
        return self

    def dequeue(self):
        if self.queue == self.empty:
            return self
        else:
            self.queue = self.queue[1:]
            return self

if __name__ == "__main__":
    q = Queue()
    print(q.enqueue(1).enqueue(2).dequeue().enqueue(3).enqueue(4).dequeue().queue)

    q2 = Queue()
    print(q2.enqueue(1).dequeue().dequeue().dequeue().queue)